substrate/pkg/controller/substrate/cluster/config.go

```diff
@@ -0,0 +1,259 @@
+	}
+	logging.FromContext(ctx).Infof("Deleted S3 bucket %s", aws.StringValue(discovery.Name(substrate)))
+	return reconcile.Result{}, nil
+}
+
+func (c *Config) generateCerts(cfg *kubeadm.InitConfiguration, substrate *v1alpha1.Substrate) error {
+	cfg.CertificatesDir = path.Join(clusterCertsBasePath, substrate.Name, certPKIPath)
+	certTree, err := certsphase.GetDefaultCertList().AsMap().CertTree()
+	if err != nil {
+		return err
+	}
+	if err := certTree.CreateTree(cfg); err != nil {
+		return fmt.Errorf("error creating cert tree, %w", err)
+	}
+	return nil
+}
+
+func (c *Config) generateStaticPodManifests(cfg *kubeadm.InitConfiguration, substrate *v1alpha1.Substrate) error {
+	manifestDir := path.Join(clusterCertsBasePath, substrate.Name, clusterManifestPath)
+	// etcd phase adds cfg.CertificatesDir to static pod yaml for pods to read the certs from
+	cfg.CertificatesDir = certPKIPath
+	if err := etcdphase.CreateLocalEtcdStaticPodManifestFile(
+		manifestDir, "", cfg.NodeRegistration.Name, &cfg.ClusterConfiguration, &cfg.LocalAPIEndpoint, false); err != nil {
+		return fmt.Errorf("error creating local etcd static pod manifest file %w", err)
+	}
+	for _, componentName := range []string{
+		kubeadmconstants.KubeAPIServer,
+		kubeadmconstants.KubeControllerManager,
+		kubeadmconstants.KubeScheduler} {
+		err := controlplane.CreateStaticPodFiles(path.Join(clusterCertsBasePath, substrate.Name, clusterManifestPath), "",
+			&cfg.ClusterConfiguration, &cfg.LocalAPIEndpoint, false, componentName)
+		if err != nil {
+			return fmt.Errorf("creating static pod file for %v, %w", componentName, err)
+		}
+	}
+	return nil
+}
+
+func (c *Config) ensureS3Bucket(ctx context.Context, substrate *v1alpha1.Substrate) error {
+	if _, err := c.S3.CreateBucket(&s3.CreateBucketInput{
+		Bucket: discovery.Name(substrate),
+		CreateBucketConfiguration: &s3.CreateBucketConfiguration{
+			LocationConstraint: c.S3.Config.Region,
+		},
+	}); err != nil && err.(awserr.Error).Code() != s3.ErrCodeBucketAlreadyOwnedByYou {
+		return fmt.Errorf("creating S3 bucket, %w", err)
+	}
+	logging.FromContext(ctx).Infof("Ensured s3 bucket %s", aws.StringValue(discovery.Name(substrate)))
+	return nil
+}
+
+func (c *Config) uploadDirectories(ctx context.Context, substrate *v1alpha1.Substrate) error {
+	if err := c.S3Uploader.UploadWithIterator(ctx, NewDirectoryIterator(
+		aws.StringValue(discovery.Name(substrate)), localConfigPath(substrate))); err != nil {
+		return fmt.Errorf("uploading to S3 %w", err)
+	}
+	logging.FromContext(ctx).Infof("Uploaded configs from %q to bucket %q",
+		localConfigPath(substrate), aws.StringValue(discovery.Name(substrate)))
+	return nil
+}
+
+func localConfigPath(substrate *v1alpha1.Substrate) string {
+	return path.Join(clusterCertsBasePath, substrate.Name)
+}
+
+func defaultClusterConfig(substrate *v1alpha1.Substrate) (*kubeadm.InitConfiguration, error) {
+	defaultStaticConfig, err := configutil.DefaultedStaticInitConfiguration()
```

Comment:
Might be nice to runtime.Must() here. Hard to imagine this can actually error at runtime.